wpilibc/src/main/native/include/frc/trajectory/TrapezoidProfile.h

```diff
@@ -66,7 +77,51 @@ class TrapezoidProfile {
    * @param initial     The initial state (usually the current state).
    */
   TrapezoidProfile(Constraints constraints, State goal,
-                   State initial = State{0_m, 0_mps});
+                   State initial = State{Distance_t(0), Velocity_t(0)})
+      : m_direction{ShouldFlipAcceleration(initial, goal) ? -1 : 1},
+        m_constraints(constraints),
+        m_initial(Direct(initial)),
+        m_goal(Direct(goal)) {
+    if (m_initial.velocity > m_constraints.maxVelocity) {
+      m_initial.velocity = m_constraints.maxVelocity;
+    }
+
+    // Deal with a possibly truncated motion profile (with nonzero initial or
+    // final velocity) by calculating the parameters as if the profile began and
+    // ended at zero velocity
+    units::second_t cutoffBegin =
+        m_initial.velocity / m_constraints.maxAcceleration;
+    Distance_t cutoffDistBegin =
+        cutoffBegin * cutoffBegin * m_constraints.maxAcceleration / 2.0;
+
+    units::second_t cutoffEnd = m_goal.velocity / m_constraints.maxAcceleration;
+    Distance_t cutoffDistEnd =
+        cutoffEnd * cutoffEnd * m_constraints.maxAcceleration / 2.0;
+
+    // Now we can calculate the parameters as if it was a full trapezoid instead
+    // of a truncated one
+
+    Distance_t fullTrapezoidDist = cutoffDistBegin +
+                                   (m_goal.position - m_initial.position) +
+                                   cutoffDistEnd;
+    units::second_t accelerationTime =
+        m_constraints.maxVelocity / m_constraints.maxAcceleration;
+
+    Distance_t fullSpeedDist =
+        fullTrapezoidDist -
+        accelerationTime * accelerationTime * m_constraints.maxAcceleration;
+
+    // Handle the case where the profile never reaches full speed
+    if (fullSpeedDist < Distance_t(0)) {
+      accelerationTime =
+          units::math::sqrt(fullTrapezoidDist / m_constraints.maxAcceleration);
+      fullSpeedDist = Distance_t(0);
+    }
+
+    m_endAccel = accelerationTime - cutoffBegin;
+    m_endFullSpeed = m_endAccel + fullSpeedDist / m_constraints.maxVelocity;
+    m_endDeccel = m_endFullSpeed + accelerationTime - cutoffEnd;
+  }
```

Comment:
These implementations are long enough they should be probably moved to a .inl file.